File: src/main/java/com/github/wgzhao/dbquery/util/DbUtil.java
package com.github.wgzhao.dbquery.util;

import com.github.wgzhao.dbquery.dto.CommResponse;
import com.github.wgzhao.dbquery.entities.DataSources;
import com.github.wgzhao.dbquery.dto.DatabaseType;
import lombok.extern.slf4j.Slf4j;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

@Slf4j
public class DbUtil
{

    // guess database type from jdbc url
    public static DatabaseType getDbType(String url)
    {
        if (url.contains("mysql")) {
            return DatabaseType.MYSQL;
        }
        else if (url.contains("postgresql")) {
            return DatabaseType.POSTGRESQL;
        }
        else if (url.contains("sqlserver")) {
            return DatabaseType.SQLSERVER;
        }
        else if (url.contains("oracle")) {
            return DatabaseType.ORACLE;
        }
        else if (url.contains("db2")) {
            return DatabaseType.DB2;
        }
        else if (url.contains("h2")) {
            return DatabaseType.H2;
        }
        else if (url.contains("hsqldb")) {
            return DatabaseType.HSQLDB;
        }
        else if (url.contains("sqlite")) {
            return DatabaseType.SQLITE;
        }
        else if (url.contains("mariadb")) {
            return DatabaseType.MARIADB;
        }
        else {
            return DatabaseType.OTHER;
        }
    }

    // test connection for jdbc url
    public static CommResponse testConnect(DataSources db)
    {
        Map<String, Object> result = new HashMap<>();
        try {
            Class.forName(db.getDriver());
            Properties properties = new Properties();
            properties.setProperty("user", db.getUsername());
            if (db.getPassword() != null) {
                properties.setProperty("password", db.getPassword());
            }
            // set connect timeout for 5 seconds
            if (! db.getUrl().contains("trino")) {
                properties.setProperty("connectTimeout", "5000");
            }
            Connection connection = DriverManager.getConnection(db.getUrl(), properties);
            if (connection != null) {
                if (db.getUrl().contains("oracle")) {
                    connection.createStatement().execute("SELECT 1 FROM DUAL");
                }
                else {
                    connection.createStatement().execute("SELECT 1");
                }
                connection.close();
                return new CommResponse(true, "");
            }
        }
        catch (ClassNotFoundException e) {
            log.error("Driver not found: {}", db.getDriver());
            return new CommResponse(false, "Driver not found: " + db.getDriver());
        }
        catch (SQLException e) {
            log.error("Connection failed: {}", e.getMessage());
            return new CommResponse(false, e.getMessage());
        }
        return new CommResponse(false, "Unknown error");
    }
}

File: src/main/kotlin/com/github/wgzhao/dbquery/util/DbUtil.kt
package com.github.wgzhao.dbquery.util

import com.github.wgzhao.dbquery.dto.DatabaseType
import com.github.wgzhao.dbquery.entities.DataSources
import org.slf4j.LoggerFactory
import java.sql.DriverManager
import java.sql.SQLException
import java.util.Properties

object DbUtil {
    private val logger = LoggerFactory.getLogger(DbUtil::class.java)

    // guess database type from jdbc url
    fun getDbType(url: String): DatabaseType = when {
        url.contains("mysql", ignoreCase = true) -> DatabaseType.MYSQL
        url.contains("postgresql", ignoreCase = true) -> DatabaseType.POSTGRESQL
        url.contains("sqlserver", ignoreCase = true) -> DatabaseType.SQLSERVER
        url.contains("oracle", ignoreCase = true) -> DatabaseType.ORACLE
        url.contains("db2", ignoreCase = true) -> DatabaseType.DB2
        url.contains("h2", ignoreCase = true) -> DatabaseType.H2
        url.contains("hsqldb", ignoreCase = true) -> DatabaseType.HSQLDB
        url.contains("sqlite", ignoreCase = true) -> DatabaseType.SQLITE
        url.contains("mariadb", ignoreCase = true) -> DatabaseType.MARIADB
        else -> DatabaseType.OTHER
    }

    // test connection for jdbc url
    fun testConnect(db: DataSources): String {
        try {
            Class.forName(db.driver)
            val urlLower = db.url.lowercase()
            val properties = Properties().apply {
                setProperty("user", db.username)
                db.password?.let { setProperty("password", it) }
                if (!urlLower.contains("trino")) {
                    setProperty("connectTimeout", "5000")
                }
            }

            DriverManager.getConnection(db.url, properties).use { connection ->
                connection.createStatement().use { stmt ->
                    if (urlLower.contains("oracle")) {
                        stmt.execute("SELECT 1 FROM DUAL")
                    } else {
                        stmt.execute("SELECT 1")
                    }
                }
                return "success"
            }
        } catch (e: ClassNotFoundException) {
            logger.error("Driver not found: {}", db.driver, e)
            return "Driver not found: ${db.driver}"
        } catch (e: SQLException) {
            logger.error("Connection failed: {}", e.message, e)
            return e.message ?: e.toString()
        }
    }
}
